### ISSUE TITLE
-Wshift-overflow in drivers/net/ethernet/qlogic/qla3xxx.c

### ISSUE THREAD

**Issue by tpimh** [ClangBuiltLinux/linux#127]
```
drivers/net/ethernet/qlogic/qla3xxx.c:384:24: warning: signed shift result (0xF00000000)
      requires 37 bits to represent, but 'int' only has 32 bits [-Wshift-overflow]
                           ((ISP_NVRAM_MASK << 16) | qdev->eeprom_cmd_data));
                             ~~~~~~~~~~~~~~ ^  ~~
1 warning generated.
```

**Comment by nathanchance** [comment]
This is the last one in the "low hanging fruit" category (in my opinion at least).

Full function:

```C
/*
 * Caller holds hw_lock.
 */
static void fm93c56a_select(struct ql3_adapter *qdev)
{
	struct ql3xxx_port_registers __iomem *port_regs =
			qdev->mem_map_registers;
	__iomem u32 *spir = &port_regs->CommonRegs.serialPortInterfaceReg;

	qdev->eeprom_cmd_data = AUBURN_EEPROM_CS_1;
	ql_write_nvram_reg(qdev, spir, ISP_NVRAM_MASK | qdev->eeprom_cmd_data);
	ql_write_nvram_reg(qdev, spir,
			   ((ISP_NVRAM_MASK << 16) | qdev->eeprom_cmd_data));
}
```

My guess is that the second statement shouldn't be there since `ISP_NVRAM_MASK` is defined as `0x000F << 16` in `drivers/net/ethernet/qlogic/qla3xxx.h` and it's the only enum value shifted. That would also match `fm93c56a_deselect` further down.

```C
/*
 * Caller holds hw_lock.
 */
static void fm93c56a_deselect(struct ql3_adapter *qdev)
{
	struct ql3xxx_port_registers __iomem *port_regs =
			qdev->mem_map_registers;
	__iomem u32 *spir = &port_regs->CommonRegs.serialPortInterfaceReg;

	qdev->eeprom_cmd_data = AUBURN_EEPROM_CS_0;
	ql_write_nvram_reg(qdev, spir, ISP_NVRAM_MASK | qdev->eeprom_cmd_data);
}
```

I'll reach out to the maintainers tomorrow for a comment, this has been like this since the driver was first introduced back in 2006 (5a4faa873782d748960b02fdec95e3d4d2e3ae38) so it's either intentional, magically worked on overflowing, or broken lol.

Suggested solution:

```diff
diff --git a/drivers/net/ethernet/qlogic/qla3xxx.c b/drivers/net/ethernet/qlogic/qla3xxx.c
index b48f76182049..10b075bc5959 100644
--- a/drivers/net/ethernet/qlogic/qla3xxx.c
+++ b/drivers/net/ethernet/qlogic/qla3xxx.c
@@ -380,8 +380,6 @@ static void fm93c56a_select(struct ql3_adapter *qdev)
 
        qdev->eeprom_cmd_data = AUBURN_EEPROM_CS_1;
        ql_write_nvram_reg(qdev, spir, ISP_NVRAM_MASK | qdev->eeprom_cmd_data);
-       ql_write_nvram_reg(qdev, spir,   
-                          ((ISP_NVRAM_MASK << 16) | qdev->eeprom_cmd_data));
 }
 
 /*
```

**Comment by nathanchance** [comment]
Actually inquiry sent now: https://lore.kernel.org/lkml/20181004070558.GA6535@flashbox/

**Comment by nathanchance** [comment]
Pending patch, depending on what response I get from upstream: https://github.com/nathanchance/patches/blob/922561bc18ae/linux/pending/0003-net-qla3xxx-Remove-overflowing-shift-statement.patch

**Comment by nathanchance** [comment]
Decided to send the patch: https://lore.kernel.org/lkml/20181013021457.13608-1-natechancellor@gmail.com/

**Comment by nickdesaulniers** [comment]
Writing two different values to the same register seems odd (unless the hardware expects that, seems kind of dump; why not write a combined value once).  Unless they meant to write to two different registers via two different write calls?

**Comment by nathanchance** [comment]
Patch accepted: https://git.kernel.org/pub/scm/linux/kernel/git/davem/net.git/commit/?id=8c3bf9b62b667456a57aefcf1689e826df146159

Guess we'll see if it breaks the hardware since none of the Cavium guys replied: https://lore.kernel.org/lkml/20181017.215322.873219355646098517.davem@davemloft.net/

**Comment by nathanchance** [comment]
Patch merged into mainline, will be in final 4.19: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=8c3bf9b62b667456a57aefcf1689e826df146159
